[WIP] Outlier detection

renesass opened the pull request.

renesass committed:
Implemented outlier detection benchmarks

renesass committed:
Implemented outlier detection benchmarks

renesass committed:
Implemented outlier detection benchmarks

renesass committed:
Implemented outlier detection benchmarks

renesass committed:
Added container data

renesass committed:
Added container data

renesass committed:
Added container data

renesass committed:
Added container data

renesass committed:
Updated classnames and filenames

renesass committed:
Updated classnames and filenames

renesass committed:
Updated classnames and filenames

renesass committed:
Updated classnames and filenames

renesass committed:
Changed placeholder examples (#110)

renesass committed:
Added more outlier datasets

renesass committed:
Added more outlier datasets

renesass committed:
Added more outlier datasets

renesass committed:
Added more outlier datasets

renesass committed:
Improved instructions (#110)

renesass committed:
Replaced automl github name

renesass committed:
Added container information (#110)

renesass committed:
Changed container name

renesass committed:
Changed container name

renesass committed:
Changed container name

renesass committed:
Changed container name

renesass committed:
Try to get container running

renesass committed:
Try to get container running

renesass committed:
Try to get container running

renesass committed:
Try to get container running

renesass committed:
Try to get container running

renesass committed:
Try to get container running

renesass committed:
Changed testing container locally (#110)

renesass committed:
Simplifications (#110)

PhMueller committed:
Add startup add connect client (#109)

* Split the contaier client interface into mulitple steps.

By doing this, we can have a container running in the background, and afterwards create a seperate proxy connection to it, without going through its init-phase. (basically, we just connect to it.). 
This PR also adds an example that shows how to use the new interface correctly.

renesass committed:
Improvements (#110)

renesass committed:
More improvements (#110)

KEggensperger commented on extra_requirements/outlier_detection.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "outlier_detection": ["torch==1.9.0", "pytorch_lightning==1.3.8", "scikit-learn==0.24.2", "scipy==1.7.0"]
```
Is there a specific feature you need from scipy=1.7.0?
Scipy>=1.4.1 is already installed for HPOBench, see [here](https://github.com/automl/HPOBench/blob/master/requirements.txt) (and probably scikit-learn and pytorch will require it as well). Would it be possible to remove it from this list?

renesass replied:
removed it

KEggensperger commented on hpobench/benchmarks/od/backbones/mlp.py:
```diff
@@ -0,0 +1,159 @@
+import torch
```
This is not directly benchmark code. Could you move all code that doe not contain classes inheriting from AbstractBenchmark to `od_utils/`?

KEggensperger replied:
Same for the other files below

renesass replied:
I created a new folder hpobench/dependencies/od.
Makes sense to have such a folder also for other benchmarks in the future. Let me know what you think.

PhMueller replied:
Fine by me. I have placed it for the TurBO-Benchmarks into `utils/benchmarks/....`. 
But we can also call it dependencies.

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+        """
+        Parameters
+        ----------
+        dataset_name : str
```
Please add a list with allowed dataset names

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+                           rng: Union[np.random.RandomState, int, None] = None, **kwargs) -> Dict:
+        """
+        Trains the autoencoder with 4-fold cross-validation.
+        Training ends if validation AUPR is not getting better after 10 epochs.
```
Is this "area under the PR curve", could you please add that? Also, the benchmark should return something to be minimized, so this is probably misleading?

renesass replied:
it returns something to be minimized, just didn't mention it properly in the description.

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+        Returns
+        -------
+        Dict -
+            function_value : validation AUPR
```
Same here, is this "1- mean validation AUPR"?

renesass replied:
you are completely right. I changed all of that. Sorry that I didn't see that in the first place.

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+            function_value : validation AUPR
+            cost : time to train and evaluate the model
+            info : Dict
+                train_loss : training loss
```
And this is then "1-X" as well or is it the actual aupr value (or even the real loss value)?

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+
+class ODAutoencoder(AbstractBenchmark):
+    """
+    Hyperparameter optimization task to optimize an autoencoder
```
This would be the right location for some more details, e.g. what you sent me:

A fully connected autoencoder implemented in PyTorch Lightning~\cite{falcon2019pytorch} with a maximum of four layers (including latent layer) is optimized on AUPR. In addition to the neural architecture hyperparameters, the benchmark includes multiple regularization techniques, scalers, and activation functions to train the autoencoder on 15 different Outlier Detection DataSets~\cite{rayana2016}\footnote{annthyroid, arrhytmia, breastw, cardio, ionosphere, mammography, musk, optdigits, pendigits, pima, satellite, satimage-2, thyroid, vowels, wbc}. Cross-validation, early stopping, and the number of training epochs as fidelity complete the benchmark.

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+            # Get epoch with the highest validation aupr
+            index = np.argmax(np.array(model.val_auprs))
+
+            train_loss = 1 - model.train_losses[index]
```
Please double check this.

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+                ],
+            )
+
+            trainer.fit(
```
Can this fail, e.g. not converge? If yes, then this should be surrounded by a try/catch and return 1 (or whatever is the highest for your metric)

renesass replied:
good point. thanks for pointing out!

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+            'function_value': float(test_aupr),
+            'cost': cost,
+            'info': {
+                'train_loss': float(np.mean(np.array(model.train_losses))),
```
This also wants to be double checked to be consistent with above (here, you don't do 1-loss)

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+        skip_connection = CSH.CategoricalHyperparameter('skip_connection', choices=[True, False], default_value=False)
+        batch_normalization = CSH.CategoricalHyperparameter('batch_normalization', choices=[True, False], default_value=False)
+        dropout = CSH.CategoricalHyperparameter('dropout', choices=[True, False], default_value=True)
+        dropout_rate = CSH.UniformFloatHyperparameter('dropout_rate', 0, 0.8, default_value=0.5)
```
So, theorethically dropout can be turned off by setting it to 0 although it is turned on?

renesass replied:
set it to 0.2

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+            #],
+        }
+
+        choice = CSH.CategoricalHyperparameter(
```
If there's only one optimizer, then this hyperparameter is constant and not necessary.

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+        )
+        cs.add_hyperparameter(choice)
+
+        for optimizer, hyperparameters in optimizers.items():
```
same here

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+            )
+
+        # Batch size
+        batch_size = CSH.UniformIntegerHyperparameter('batch_size', lower=16, upper=512, default_value=128)
```
looks like it should be on a log scale?

renesass replied:
why should batch_size be on log scale?

KEggensperger replied:
I guess, it doesn't have to. I believe I've seen it more often optimized on a logscale. If you've optimized it like that for your work, it's fine by me.

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+        fidel_space = CS.ConfigurationSpace(seed=seed)
+
+        fidel_space.add_hyperparameters([
+            CS.UniformIntegerHyperparameter("epochs", lower=10, upper=100, default_value=10, log=False),
```
default should be the highest value, i.e. 100

KEggensperger commented on hpobench/benchmarks/od/od_ae.py:
```diff
@@ -0,0 +1,426 @@
+
+        return {
+            'name': self.get_name(),
+            'references': [
```
Please add what this reference is for

renesass replied:
what do you mean?

KEggensperger commented on hpobench/benchmarks/od/od_kde.py:
```diff
@@ -0,0 +1,94 @@
+class ODKernelDensityEstimation(ODTraditional):
+    """
+    Hyperparameter optimization task to optimize OCSVM
+    for outlier detection.
```
Also here, some more details are necessary for someone new to understand what this benchmark is doing

KEggensperger commented on hpobench/benchmarks/od/trad_benchmark.py:
```diff
@@ -0,0 +1,262 @@
+        Returns
+        -------
+        Dict -
+            function_value : validation loss
```
Please also make sure that this is described correctly and returns the correct value

KEggensperger commented on hpobench/benchmarks/od/od_ocsvm.py:
```diff
@@ -0,0 +1,88 @@
+        hp_gamma = float(configuration['gamma'])
+        hp_nu = float(configuration['nu'])
+
+        return OneClassSVM(kernel="rbf", gamma=hp_gamma, nu=hp_nu)
```
is there a way to set the seed?

KEggensperger replied:
same question for all models

renesass replied:
Unfortunately not. Neither OCSVM nor KDE provides an option for a seed.

KEggensperger commented on hpobench/benchmarks/od/trad_benchmark.py:
```diff
@@ -0,0 +1,262 @@
+            Configuration for the SVM Model
+        fidelity: Dict, None
+            Fidelity parameters, check get_fidelity_space(). Uses default (max) value if None.
+        shuffle : bool
```
this is not imlemented and should be removed

KEggensperger commented on hpobench/benchmarks/od/trad_benchmark.py:
```diff
@@ -0,0 +1,262 @@
+        model.fit(X_train, y_train)
+
+        # Compute validation error
+        train_loss = 1 - self.calculate_aupr(model, X_train, y_train)
```
Also, shouldn't this be the same code as for the network solution?

renesass replied:
the network is trained on the reconstruction distance. However, I added train_aupr too.

renesass replied:
I removed it again. Since the training data are mostly clean datapoints, it makes no sense to calculate the train aupr.

renesass replied:
I added it again and set the contamination ratio to 10%.

KEggensperger commented on hpobench/benchmarks/od/trad_benchmark.py:
```diff
@@ -0,0 +1,262 @@
+
+        Fidelities
+        ----------
+        dataset_fraction: float - [0.1, 1]
```
not implemented

KEggensperger commented on hpobench/container/recipes/od/Singularity.ODAutoencoder:
```diff
@@ -0,0 +1,23 @@
+    && cd /home \
+    && git clone https://github.com/automl/HPOBench.git \
+    && cd HPOBench \
+    && git checkout master \
```
development for now

KEggensperger commented on hpobench/container/recipes/od/Singularity.ODKernelDensityEstimation:
```diff
@@ -0,0 +1,23 @@
+    && cd /home \
+    && git clone https://github.com/automl/HPOBench.git \
+    && cd HPOBench \
+    && git checkout master \
```
development for now

KEggensperger commented on hpobench/container/recipes/od/Singularity.ODOneClassSupportVectorMachine:
```diff
@@ -0,0 +1,23 @@
+    && cd /home \
+    && git clone https://github.com/automl/HPOBench.git \
+    && cd HPOBench \
+    && git checkout OutlierDetection \
```
development for now

KEggensperger commented on hpobench/util/od_data_manager.py:
```diff
@@ -0,0 +1,348 @@
+
+    Parameters
+    ----------
+    openml_task_id : int
```
openml is not involved here?

KEggensperger reviewed (changes requested):
Made a first pass. Looks good, but the objective values need to be double-checked and be better described. Thank already.

renesass committed:
Fixed requests (#111)

renesass committed:
Fixed requests (#111)

renesass committed:
Fixed requests (#111)

renesass committed:
Fixed requests (#111)

PhMueller committed:
Fix Inf Values in PYBNN Benchmark (#114)

* Fix a bug in the pybnn benchmarks.
  Limit the variance before computing the nll. (inf otherwise)

PhMueller committed:
Prepare new release (#115)

renesass committed:
Changed ODDS links

renesass committed:
Changed ODDS links

renesass committed:
Changed ODDS links

Commit:
Update __version__.py (#117)

* Update __version__.py

* Update changelog.md

KEggensperger reviewed (approved):
Looks good to me, but I'll leave it up to @PhMueller to merge this eventually.

PhMueller committed:
TowardsExperiments (#118)

* User correct container version for SVMSurrogate and ParamNet

PhMueller committed:
Remove Logger statement

When Trainer receives `None` as input for the logger it will use the default logger.
Those default logger will then write to non-mounted directories.
This causes a crash when using the benchmark inside of a container.

PhMueller committed:
Remove Logger statement

When Trainer receives `None` as input for the logger it will use the default logger.
Those default logger will then write to non-mounted directories.
This causes a crash when using the benchmark inside of a container.

PhMueller committed:
Remove Logger statement

When Trainer receives `None` as input for the logger it will use the default logger.
Those default logger will then write to non-mounted directories.
This causes a crash when using the benchmark inside of a container.

PhMueller committed:
Remove Logger statement

When Trainer receives `None` as input for the logger it will use the default logger.
Those default logger will then write to non-mounted directories.
This causes a crash when using the benchmark inside of a container.

PhMueller committed:
UPDATE Container interface

Merge the 3 files and also add a entry point for the 3 benchmarks.
Then, we can use a single container for all the 3 benchmarks.

PhMueller committed:
UPDATE Container interface

Merge the 3 files and also add a entry point for the 3 benchmarks.
Then, we can use a single container for all the 3 benchmarks.

PhMueller committed:
UPDATE Container interface

Merge the 3 files and also add a entry point for the 3 benchmarks.
Then, we can use a single container for all the 3 benchmarks.

PhMueller committed:
UPDATE Container interface

Merge the 3 files and also add a entry point for the 3 benchmarks.
Then, we can use a single container for all the 3 benchmarks.

PhMueller committed:
Update Container Interface (2): merge recipes

PhMueller committed:
Update Container Interface (2): merge recipes

PhMueller committed:
Update Container Interface (2): merge recipes

PhMueller committed:
Update Container Interface (2): merge recipes

PhMueller committed:
Set seeds

PhMueller committed:
Set seeds

PhMueller committed:
Set seeds

PhMueller committed:
Set seeds

PhMueller committed:
REFACTOR + PEP8 + FLAKE8

PhMueller committed:
REFACTOR + PEP8 + FLAKE8

PhMueller committed:
REFACTOR + PEP8 + FLAKE8

PhMueller committed:
REFACTOR + PEP8 + FLAKE8

PhMueller committed:
Merge remote-tracking branch 'origin/OutlierDetection' into OutlierDetection

PhMueller reviewed (approved):
I will upload the benchmark container to the registry.

PhMueller committed:
Add missing __init__ files.

PhMueller committed:
Add missing __init__ files.

PhMueller committed:
Use the container for the tests

PhMueller committed:
Use the container for the tests

PhMueller committed:
Merge remote-tracking branch 'origin/OutlierDetection' into OutlierDetection
